trie/triedb/pathdb/nodebufferlist.go

```diff
@@ -110,31 +115,196 @@ func newNodeBufferList(
 		}
 	}
 	base := newMultiDifflayer(limit, size, common.Hash{}, nodes, layers)
-	ele := newMultiDifflayer(limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
-	nf := &nodebufferlist{
+
+	var nf *nodebufferlist
+	if recovery {
+		nf = newNodeBufferListForRecovery(db, freezer, base, limit, wpBlocks, rsevMdNum, dlInMd)
+	} else {
+		ele := newMultiDifflayer(limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
+		nf = &nodebufferlist{
+			db:              db,
+			wpBlocks:        wpBlocks,
+			rsevMdNum:       rsevMdNum,
+			dlInMd:          dlInMd,
+			limit:           limit,
+			base:            base,
+			head:            ele,
+			tail:            ele,
+			count:           1,
+			persistID:       rawdb.ReadPersistentStateID(db),
+			stopCh:          make(chan struct{}),
+			waitStopCh:      make(chan struct{}),
+			forceKeepCh:     make(chan struct{}),
+			waitForceKeepCh: make(chan struct{}),
+			keepFunc:        keepFunc,
+		}
+	}
+
+	go nf.loop()
+
+	log.Info("new node buffer list", "proposed block interval", nf.wpBlocks,
+		"reserve multi difflayers", nf.rsevMdNum, "difflayers in multidifflayer", nf.dlInMd,
+		"limit", common.StorageSize(limit), "layers", layers, "persist id", nf.persistID, "base_size", size)
+	return nf
+}
+
+// newNodeBufferListForRecovery initializes the node buffer list for recovering diff layers
+func newNodeBufferListForRecovery(db ethdb.Database, freezer *rawdb.ResettableFreezer, base *multiDifflayer,
```

Comment:
Maybe  replace `log.Crit` to `return error`, if `newNodeBufferListForRecovery` return error, `loadLayers` will goto 280 lines, to start form no recovery mode.

Reply:
I think it should be exited if `failed to recover node buffer list`.